src/firestore/indexes.ts

```diff
@@ -0,0 +1,560 @@
+   * @param indexes an array of objects, each will be validated and then converted
+   * to an {@link IndexSpecEntry}.
+   */
+  async deploy(project: string, indexes: any[], fieldOverrides: any[]): Promise<any> {
+    const spec = this.upgradeOldSpec({
+      indexes,
+      fieldOverrides,
+    });
+
+    this.validateSpec(spec);
+
+    // Now that the spec is validated we can safely assert these types.
+    const indexesToDeploy: Spec.Index[] = spec.indexes;
+    const fieldOverridesToDeploy: Spec.FieldOverride[] = spec.fieldOverrides;
+
+    const existingIndexes = await this.listIndexes(project);
+    const existingFieldOverrides = await this.listFieldOverrides(project);
+
+    if (existingIndexes.length > indexesToDeploy.length) {
+      utils.logBullet(
+        "There are some indexes defined in your project that are not present in your " +
+          "firestore indexes file. Run firebase firestore:indexes and save the result to correct the discrepancy."
+      );
+    }
+
+    indexesToDeploy.forEach(async (index) => {
+      const exists = existingIndexes.some((x) => this.indexMatchesSpec(x, index));
+      if (exists) {
+        logger.debug(`Skipping existing index: ${JSON.stringify(index)}`);
+        return;
+      }
+
+      logger.debug(`Creating new index: ${JSON.stringify(index)}`);
+      await this.createIndex(project, index);
+    });
+
+    if (existingFieldOverrides.length > fieldOverridesToDeploy.length) {
+      utils.logBullet(
+        "There are some field overrides defined in your project that are not present in your " +
+          "firestore indexes file. Run firebase firestore:indexes and save the result to correct the discrepancy."
+      );
+    }
+
+    fieldOverridesToDeploy.forEach(async (field) => {
+      const exists = existingFieldOverrides.some((x) => this.fieldMatchesSpec(x, field));
+      if (exists) {
+        logger.debug(`Skipping existing field override: ${JSON.stringify(field)}`);
+        return;
+      }
+
+      logger.debug(`Updating field override: ${JSON.stringify(field)}`);
+      await this.patchField(project, field);
+    });
+  }
+
+  /**
+   * List all indexes that exist on a given project.
+   * @param project the Firebase project id.
+   */
+  async listIndexes(project: string): Promise<API.Index[]> {
+    const url = `projects/${project}/databases/(default)/collectionGroups/-/indexes`;
+
+    const res = await api.request("GET", `/v1beta2/${url}`, {
+      auth: true,
+      origin: api.firestoreOrigin,
+    });
+
+    const indexes = res.body.indexes;
+    return indexes.map((index: any) => {
+      // Ignore any fields that point at the document ID, as those are implied
+      // in all indexes.
+      const fields = index.fields.filter((field: API.IndexField) => {
+        return field.fieldPath !== "__name__";
+      });
+
+      return {
+        name: index.name,
+        state: index.state,
+        queryScope: index.queryScope,
+        fields,
+      } as API.Index;
```

Comment:
can `API.Index` be placed in the function signature and then removed from here? Does that work?

```typescript
.map((index: any): API.Index => {
  //...
  return {};
});